doc/developer/design/20210311_prometheus_source.md

```diff
@@ -0,0 +1,141 @@
+# Summary
+
+This design doc introduces a way to import data from our internal prometheus registry into materialized views that can be inspected with SQL commands.
+
+# Goals
+
+* Allow our users to read & make decisions based on our metrics data without setting up a whole metrics ingestion/presentation pipeline.
+
+* Let administrators inspect that metrics data (as much as metrics' attached data allows) together with the other available data we present in system tables.
+
+* Retain a long-enough history of metrics values to allow making decisions based on how the metrics values change (but short enough to not consume unbounded memory).
+
+# Non-Goals
+
+None of the following are intended to be done with this design:
+
+* Export our system table logs as prometheus (yet) -- this design covers only the "metrics import" direction.
+
+* Expose any `CREATE SOURCE` mechanics for importing from arbitrary/external prometheus endpoints.
+
+* Come up with a data retention scheme other than "older than n minutes gets dropped". If this was a full-featured time series database design, we'd want to thin metrics into minute/hour/day buckets. This isn't that.
+
+# Description
+
+<!--
+    Describe the approach in detail. If there is no clear frontrunner, feel free to list all approaches in alternatives.
+    If applicable, be sure to call out any new testing/validation that will be required
+-->
+
+By implementing this design, we'll make the following changes to materialized:
+
+1. Introduce a background task that regularly scrapes the internal metrics registry using the registry's [`gather`](https://docs.rs/prometheus/0.12.0/prometheus/struct.Registry.html#method.gather) method (currently we can use the global registry, but later we'll need a handle to the ["materialized" metrics registry][#5825]; the mechanism will remain the same).
+2. Ingest the generated metrics into a dataflow that populates metrics into tables like the following:
+
+```sql
+CREATE TABLE metrics_table_name (
+    metric TEXT,
+    ts TIMESTAMP,
+    labels JSONB,
+    value float8
+);
+
+CREATE TABLE metrics_table_meta (
+    metric TEXT,
+    metric_type TEXT,
+    help TEXT
+);
+```
+
+3. This not-quite-source polls the endpoint once a minute (via the configured the `poll_interval` duration), and retains metric readings for a number of minutes (strawman proposal: 5 minutes) to make it possible for users to visualize changes in the metrics as they would with a "full" metrics pipeline.
```

Comment:
I think a default poll interval of 1 or 5 seconds is more likely to be pleasant hopefully without meaningfully impacting performance of the system as a whole.

Reply:
I was going with the [default prometheus poll interval](https://prometheus.io/docs/prometheus/latest/configuration/configuration/), 60s. Do we configure our metric pipeline for shorter intervals? Then this definitely ought to change.

Reply:
Yeah our cloud prometheus polls typically around every 15 seconds,b ut I don't know that that's necessarily the right metaphor for what this data is meant to reflect. The cloud prometheus polling interval needs to respect the fact that a single prometheus node is often responsible for scraping tens to hundreds or thousands of instances, and the data is meant to be pretty coarse.

This internal item is meant to accurately reflect what is going on internally in our system, and should generally be able to scrape in a couple milliseconds. It should generally actually probably match our [`--introspection-frequency`](https://github.com/MaterializeInc/materialize/blob/44f1e7168db258f20565363546ef5646dbaf1ba3/src/materialized/src/bin/materialized/main.rs#L94-L102) cli arg. That argument used to be named --logging-granularity or something, but the new name seems like it should affect this. (cc @benesch for his opinion)

Reply:
The open-source databases I've used update their metrics often, so +1 to updating more often. However, does INSERTing into Materialize often cause performance issues? I vaguely recall some performance limitations for tables.

Reply:
> does INSERTing into Materialize often cause performance issues

I don't expect this design to have those issues. We had some write-amplification issues when we included some of our current prometheus metrics as table metrics. Doing that meant that parsing 1 row of data from an external source could cause ~10 writes to our system tables for various counters, and it would be worse now because we've added more metrics. With the design in this doc we will increase the number of writes based on the number of metrics, but always at `--introspection-frequency` intervals, so basically we'll be inserting compacted metrics not row-by-row metrics.